include/ginkgo/core/base/lin_op.hpp

```diff
@@ -654,6 +654,20 @@ class Preconditionable {
 };
 
 
+/**
+ * A submatrix of the LinOp implementing this interface can be extracted.
+ * The row and column spans must be in range of the size of the matrix.
+ *
+ * @ingroup LinOp
+ */
+template <typename ConcreteType>
+class SubMatrixExtractable {
+public:
+    virtual std::unique_ptr<ConcreteType> create_submatrix(
+        const gko::span& row_span, const gko::span& column_span) const = 0;
+};
```

Comment:
I'm not sure about this interface - being templated on ConcreteType, it doesn't give us much flexibility in generically creating a submatrix for any matrix type. Do we expect to need this for other matrix types as well? We could for now directly add it as a Csr member function.

Reply:
I think it might make sense to have this for other formats as well. And this interface would enforce the same signatures for all the formats.

Reply:
I think if you want to be able to create submatrices in general for all matrix formats, then you either need to have this interface, or have separate functions in each matrix format. 

I am not sure we have currently strictly follow a rule where we say we only implement cheap operations for formats other than Csr and Dense. I think this would unnecessarily restrict our ability to provide functionality in other formats, and have to convert to CSR every time, which might not be reasonable.

If we are concrete matrix format agnostic (which in many solvers we are), I think we should try to maintain the same functionality and interface across all of them possibly throwing a runtime error when some functionality has not been implemented.

Regarding Dense functionality, I agree that the functionality is different there, but I think the function name, `create_submatrix` is incorrect there and we should probably change it to `create_submatrix_view` which would be more accurate for what it does. Maybe we should do this in 2.0.

Reply:
I kind of agree to both that with the one creating a few and the other extracting a submatrix, they should not be used interchangibly and that the dense function name could be more precise. I'd vote for keeping them apart for now.

Reply:
I don't think that's true? The `ValuesArray` parameter in the `Dense` constructor is taken by forwarding reference, `create` forwards, the array parameter used in `create_submatrix` is an prvalue, and the `Array` constructor turns the move-construction into a move-assignment, which moves when the executors match.